doc: support sphinx 5

bluetech opened the pull request:
Sphinx 4 (or releases since) provides two nice things (see [changelog](https://www.sphinx-doc.org/en/master/changes.html)):

- It supports "canonical" names for types. I'm not sure exactly what it is, but makes a lot more internal type annotations references resolve in the API Reference. Previously sphinx was confused by the `pytest.*` reexports from `_pytest`.
- ~~`python_use_unqualified_type_names` config which makes the API Reference signatures a little less noisy.~~

hoefling commented:
@bluetech 
> Hmm just noticed a regression of sorts - now sphinx shows a bunch of private constructors (e.g. [here](https://pytest--9248.org.readthedocs.build/en/9248/reference/reference.html#pytest.Cache) - search for `_ispytest` in the API Reference) where previously it didn't.

That's because [Sphinx added new flag for rendering typehints](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints_description_target) which is set to `"all"`, so suddenly all arguments start to be documented :-D Setting it to `autodoc_typehints_description_target = "documented"` and explicitly stating what init parameters are documented in the class docstring should help. Example: this diff

```diff
--- a/doc/en/conf.py
+++ b/doc/en/conf.py
@@ -38,6 +38,7 @@ release = ".".join(version.split(".")[:2])
 
 autodoc_member_order = "bysource"
 autodoc_typehints = "description"
+autodoc_typehints_description_target = "documented"
 todo_include_todos = 1
 
 latex_engine = "lualatex"
--- a/doc/en/reference/reference.rst
+++ b/doc/en/reference/reference.rst
@@ -347,7 +347,7 @@ Under the hood, the cache plugin uses the simple
 
 ``config.cache`` is an instance of :class:`pytest.Cache`:
 
-.. autoclass:: pytest.Cache()
+.. autoclass:: pytest.Cache(cachedir, config)
    :members:
 
 
--- a/src/_pytest/cacheprovider.py
+++ b/src/_pytest/cacheprovider.py
@@ -55,6 +55,10 @@ Signature: 8a477f597d28d172789f06886806bc55
 @final
 @attr.s(init=False, auto_attribs=True)
 class Cache:
+    """
+    :param Path cachedir:
+    :param Config config:
+    """
     _cachedir: Path = attr.ib(repr=False)
     _config: Config = attr.ib(repr=False)
```
renders
![image](https://user-images.githubusercontent.com/4455652/141188693-6fe40ab2-0167-4891-b481-d65b8f6a0b69.png)

hoefling commented:
Same is with fixtures, e.g. if I create an incomplete info field list for `capsys` via
```diff
--- a/src/_pytest/capture.py
+++ b/src/_pytest/capture.py
@@ -906,6 +906,9 @@ def capsys(request: SubRequest) -> Generator[CaptureFixture[str], None, None]:
     The captured output is made available via ``capsys.readouterr()`` method
     calls, which return a ``(out, err)`` namedtuple.
     ``out`` and ``err`` will be ``text`` objects.
+
+    :return: an instance of :class:`CaptureFixture[str] <.CaptureFixture>`.
+    :rtype: :class:`CaptureFixture[str] <.CaptureFixture>`
     """
     capman = request.config.pluginmanager.getplugin("capturemanager")
     capture_fixture = CaptureFixture[str](SysCapture, request, _ispytest=True)
```
(both fields are not needed and set just for demo) renders

![image](https://user-images.githubusercontent.com/4455652/141190154-31de51d3-9508-4f54-9904-0450440a199c.png)

What's not nice is that the info field list is part of the docstring, so the usage example is shifted down - the example could be moved to the docstring though and placed where desired.

bluetech commented:
@hoefling Thanks for looking at this.

> Setting it to autodoc_typehints_description_target = "documented" and explicitly stating what init parameters are documented in the class docstring should help.

Seems like we're going to have to do this. It would take some work -- often we omit `:param:` for "obvious" parameters/short functions.

---

Another issue with sphinx 4 I noticed is that it now always(?) renders the full signatures of overloaded functions, like `pytest.raise` and others. These were the reason we switched to `autodoc_typehints = "description"` in the first place (see #8405) so that's also something we need to fix. Maybe the right solution it to just not use autodoc for those.

hoefling commented:
> Seems like we're going to have to do this. It would take some work -- often we omit `:param:` for "obvious" parameters/short functions.

This is something I could tackle, using https://docs.pytest.org/en/stable/ as a reference.

> Another issue with sphinx 4 I noticed is that it now always(?) renders the full signatures of overloaded functions, like `pytest.raise` and others. These were the reason we switched to `autodoc_typehints = "description"` in the first place (see #8405) so that's also something we need to fix. Maybe the right solution it to just not use autodoc for those.

I think this is an issue with Sphinx 3 as well, only that Sphinx 4 now also includes full qualnames for `TypeVar`s, making a `Type[_pytest.python_api.E]` out of `Type[_E]`. I will see what I can dig out, I think [aliasing types](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_type_aliases) in Sphinx conf should mitigate that already.

bluetech commented:
Hmm Sphinx 3 no longer works in Python 3.10 (`from types import Union`) so this becomes more urgent -- I should get back to this.

The-Compiler commented:
@bluetech Anything I can help to get this in? Archlinux has [updated](https://github.com/archlinux/svntogit-packages/commit/daa634fe77bee66b804f3adc75a6672fbbf906d2) to Python 3.10 in December, and having to edit `tox.ini` every time I work on the docs is somewhat cumbersome...

hugovk commented on doc/en/requirements.txt:
```diff
@@ -2,6 +2,6 @@ pallets-sphinx-themes
 pygments-pytest>=2.2.0
 sphinx-removed-in>=0.2.0
-sphinx>=3.1,<4
+sphinx>=3.1,<5
```
Maybe bump the minimum as well? 4.2.0 is the first to support Python 3.10:

```suggestion
sphinx>=4.2.0,<5
```

And this can also be removed (need to rebase):

https://github.com/pytest-dev/pytest/blob/b4ab2f094234a0e5a1b14cf8e6d7cd9f48ecbadb/doc/en/requirements.txt#L9-L10

And should we bump Python 3.10 for `docs` and `doctesting` on the CI?

https://github.com/pytest-dev/pytest/blob/b4ab2f094234a0e5a1b14cf8e6d7cd9f48ecbadb/.github/workflows/test.yml#L145-L153

hugovk reviewed (commented):
(I meant the previous review to be a "comment")

bluetech commented:
I updated the PR. Changes:

- Now requiring Sphinx 5 instead of 4. I set the bounds to >=5, <6. Lower bound because I probably relied on something new, and upper bound to avoid possible breaking changes in Sphinx.
- I fixed the private `__init__` problem discussed above by setting `autodoc_typehints_description_target = "documented"` and painfully documenting all params. I went over all of the API reference and think it's OK, probably an improvement actually.
- I fixed the overloads problem with some hack to fool Sphinx (see commit).

So I think this is good to merge.

nicoddemus commented on src/_pytest/compat.py:
```diff
@@ -20,6 +20,16 @@
+# fmt: off
+# Workaround for https://github.com/sphinx-doc/sphinx/issues/10351.
+# If `overload` is imported from `compat` instead of from `typing`,
+# Sphinx doesn't recognize it as `overload` and the API docs for
+# overloaded functions look good again. But type checkers handle
```
IIUC, should this read:

```suggestion
# Sphinx doesn't recognize it as `overload` and the API docs for
# overloaded functions look bad. But type checkers handle
```

?

bluetech replied:
The problem is that Sphinx started recognizing overloads and rendering them, but the way it does it has some issues (see picture in https://github.com/sphinx-doc/sphinx/issues/10351, pretty similar to the we cleaned up a while back). So this hack actually makes things look good by making Sphinx *not* recognize `overload` :)

nicoddemus replied:
AHH I see, thanks!

nicoddemus commented on src/_pytest/compat.py:
```diff
@@ -20,6 +20,16 @@
+# it fine.
+# fmt: on
+if True:
+    from typing import overload as overload
```
Curious: is the "import overload as overload" part of the workaround?

bluetech replied:
This is the way to explicitly re-export a value in type checking (see https://mypy.readthedocs.io/en/stable/config_file.html?highlight=export#confval-implicit_reexport).

nicoddemus commented on src/_pytest/logging.py:
```diff
@@ -382,20 +384,19 @@ def _finalize(self) -> None:
 
-    def get_records(self, when: str) -> List[logging.LogRecord]:
+    def get_records(
+        self, when: "Literal['setup', 'call', 'teardown']"
```
No need to change, but we can use `from __future__ import annotations` and get away with writing the annotations directly (instead in string form).

bluetech replied:
Hmm it might be nice, but probably should do it for the entire code base at once?

nicoddemus replied:
I think doing piecewise is fine, and of course doing all at once is fine too. But of course, just want to comment as this is fine. 👍

nicoddemus reviewed (approved):
Looks great, looks like this was a lot of work. 😅 

Thanks a lot!

bluetech commented:
Thanks for the review @nicoddemus.
